### PR TITLE
Allow force push to protected branches

### LINKED ISSUE #22722: Allow force pushing to protected branch
### Feature Description

At this point you can only force push when a branch is not protected.

Once a branch is protected, it can be limited like this:
1. no pushes
2. all with write access can push
3. only white-listed users can push

Under option 3 (next to "Whitelist deploy keys..."), I would like an additional check box: "Allow force pushing by whitelisted users"

image

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by henrygoodman**
Fixes #22722 

### Problem
Currently, it is not possible to force push to a branch with branch protection rules in place. There are often times where this is necessary (CI workflows/administrative tasks etc). 

The current workaround is to rename/remove the branch protection, perform the force push, and then reinstate the protections.

### Solution
Provide an additional section in the branch protection rules to allow users to specify which users with push access can also force push to the branch. The default value of the rule will be set to `Disabled`, and the UI is intuitive and very similar to the `Push` section.

It is worth noting in this implementation that allowing force push does not override regular push access, and both will need to be enabled for a user to force push.

This applies to manual force push to a remote,  and also in Gitea UI updating a PR by rebase (which requires force push)

This modifies the `BranchProtection` API structs to add:
- `enable_force_push bool`
- `enable_force_push_whitelist bool`
- `force_push_whitelist_usernames string[]`
- `force_push_whitelist_teams string[]`
- `force_push_whitelist_deploy_keys bool`

### Updated Branch Protection UI:

image

### Pull Request `Update branch by Rebase` option enabled with source branch `test` being a protected branch:

image
image

### CODE REVIEW

File: options/locale/locale_en-US.ini
Comment (henrygoodman) [inline options/locale/locale_en-US.ini]: Maybe this should be changed to `Enable Force Push`

**Comment by henrygoodman** [comment]
Will be able to fix CI compliance Sunday night

**Comment by henrygoodman** [comment]
Small refactor in `f28ecbe` seemed necessary due to `make lint` showing duplicated code error, despite the 2 functions in question being quite distinct :(

**Comment by Nova-Logic** [comment]
this is very needed feature for our git workflow

**Comment by lunny** [comment]
Please fix the lint error

**Comment by Nova-Logic** [comment]
@lunny Hi! would it be possible to review this PR and include it into 1.22 release? p.s. Gitea is awesome:-)

**Comment by lunny** [comment]
> @lunny Hi! would it be possible to review this PR and include it into 1.22 release? p.s. Gitea is awesome:-)

Hi, sorry reply too late. Since the feature freeze before this PR got 2 approvals we have to move this to v1.23. We still have many work to do to make a more stable v1.22.

**Comment by lunny** [comment]
Thank you for your patience. Can you add some tests for this PR? I think there are already some for the previous protected branches in https://github.com/go-gitea/gitea/blob/fb4438a815b013895f3171f8f2f1ed22f79596de/integrations/git_test.go#L277

**Comment by henrygoodman** [comment]
Tests now extended in `doBranchProtectPRMerge` to test push capability based on existing push/force push branch protection configs. 

Let me know if this is sufficient test coverage or should extend.

**Review by lunny — APPROVED** [review]

**Comment by henrygoodman** [comment]
> Thank you for the feature! Overall it looks good to me.
> 
> Some questions (or suggestions):
> 
>     1. ~`CanForcePush` seems redundancy. Because~
>        
>        * ~If a branch is protected, then it means that it doesn't want to allow anyone to "force push"~
>        * ~It brings ambiguity, eg: what does "CanPush=false and CanForcePush=true" mean?~
>        * ~So I think it's better to drop the CanForcePush field: if a branch is protected, it could be forced push by allowed users/teams/keys~
>        * Hmm ... maybe I haven't full understood the background, will think more about it.
> 
>     2. Maybe we should use "allowlist" instead of "whitelist" from now on, because "allowlist" is widely accepted in modern days,  "whitelist" is not good in many cultures.
> 
>     3. The field names like `force_push_whitelist_user_i_ds` do not seem good enough. They are generated automatically but ideally they should be `force_push_whitelist_user_ids`.
> 
>     4. It needs a database migration

1. I agree that the combination of `CanPush=false` and `CanForcePush=true` could be confusing, as I have implemented force push as an extension of regular push permissions, therefore even if `CanForcePush=true`, it still means push will be denied since regular push is false. However, I added the disclaimer in the UI `Anyone with push access will be allowed to force push to this branch`, which could hopefully mitigate this confusion, at least in the UI (API side might still be confusing).
2. That would be good idea, however I think whitelist term is used quite a lot everywhere, it might require a new issue to general refactor.
3. Will look into if this can be revised, I agree would be cleaner
4. Will investigate migration

**Comment by wxiaoguang** [comment]
> 2. That would be good idea, however I think whitelist term is used quite a lot everywhere, it might require a new issue to general refactor.

Since this is a new feature, I think we could use the new terms, to reduce the future refactoring work (I could also help to rename them if you don't mind). In my mind after this PR we could have this:

* All terms on the UI are "allowlist" for the "protected branch" page
* The old fields are kept as-is (`WhitelistXxx`)
* The new fields use "allowlist" (`AllowlistXxx`)

**Comment by henrygoodman** [comment]
@wxiaoguang apologies this was a bit late, busy with work

Screenshot 2024-07-04 at 10 19 26 PM

In latest 2 commits:
- Updated `ProtectBranch` struct to not use the autogenerated (which converts camel case to snake case, making ID -> i_d).
- Changed all UI references in branch protection page from `whitelist` to `allowlist`
- Changed all code references for this new feature from `whitelist` to `allowlist` (existing references have not been touched)
- Performed db migration as v300 in v1.23 (let me know if this needs to be revisited, my understanding may not be correct)

**Review by wxiaoguang — APPROVED** [review]

File: models/git/protected_branch.go
Comment (wxiaoguang) [inline models/git/protected_branch.go]: Hmm, XORM struct doesn't use the `json` tag. (none of other fields has the `json` tag)

Comment (wxiaoguang) [reply models/git/protected_branch.go]: IIRC the syntax is `xorm:"approvals_whitelist_team_ids JSON TEXT"` (not quite sure, need to test)

Comment (henrygoodman) [reply models/git/protected_branch.go]: @wxiaoguang thanks for pointing out, updating in https://github.com/go-gitea/gitea/pull/28086/commits/b3b3bd497fedd0ddb1a8e8c806052912e5794657

File: models/migrations/v1_23/v300.go
Comment (wxiaoguang) [inline models/migrations/v1_23/v300.go]: The migration struct should be synchronized with the model struct.

Comment (henrygoodman) [reply models/migrations/v1_23/v300.go]: Ahh, good catch. I'll also just keep an eye on if this file needs to be updated to v301 or v302 when this needs to get merged (or if a maintainer would handle that)

Updated in https://github.com/go-gitea/gitea/pull/28086/commits/410eb5980c994167fda200bb2bab5060ef3fe9d4 (so now the migration struct fields match the model struct, hope I understood that correct)

**Review by wxiaoguang — DISMISSED** [review]

File: models/git/protected_branch.go
Comment (wxiaoguang) [inline models/git/protected_branch.go]: existing names can't be changed, otherwise it breaks exiting instances.

```suggestion
	WhitelistUserIDs              []int64 `xorm:"JSON TEXT"`
	WhitelistTeamIDs              []int64 `xorm:"JSON TEXT"`
```

File: models/git/protected_branch.go
Comment (wxiaoguang) [inline models/git/protected_branch.go]:
```suggestion
	ApprovalsWhitelistUserIDs     []int64  `xorm:"JSON TEXT"`
	ApprovalsWhitelistTeamIDs     []int64  `xorm:"JSON TEXT"`
```

File: models/git/protected_branch.go
Comment (henrygoodman) [inline models/git/protected_branch.go:554]: @wxiaoguang I thought the idea was to change the column names, since the autogenerated xorm names give `whitelist_user_i_ds`.

If we cannot modify the existing column names, I think we also have to change these back to `..._i_ds` (and the new struct fields can likely be the same for consistency sake)

Comment (wxiaoguang) [reply models/git/protected_branch.go:554]: > If we cannot modify the existing column names, I think we also have to change these back to `..._i_ds` (and the new struct fields can likely be the same for consistency sake)

Well, it looks really ugly in new code. I won't block it if most people like the name `...._i_ds`

Comment (wxiaoguang) [reply models/git/protected_branch.go:554]: Feel free to reset my changes.

Comment (henrygoodman) [reply models/git/protected_branch.go:554]: I agree it looks ugly 👍 but if we cannot modify existing columns we might be stuck with it

Comment (wxiaoguang) [reply models/git/protected_branch.go:554]: And @go-gitea/technical-oversight-committee , I am pretty busy in next months, so I wouldn't have time later.

Feel free to review & make this PR move on.

Comment (henrygoodman) [reply models/git/protected_branch.go:554]: Removed explicit xorm column names and just use `..._i_ds` as before in https://github.com/go-gitea/gitea/pull/28086/commits/a2cccd3e401617b83594ca3c0e939ee7385d0e30

**Review by wxiaoguang — APPROVED** [review]

**Comment by phild5** [comment]
I'm running 1.22.0 and i'm looking all over the place for this. Where and how do i enable it? Doesn't show up on my branch protection screen.

**Comment by henrygoodman** [comment]
> I'm running 1.22.0 and i'm looking all over the place for this. Where and how do i enable it? Doesn't show up on my branch protection screen.

 This is not integrated as part of 1.22.0, it is on the roadmap for 1.23.0 (might have to look around for an ETA). If you switch to nightly build it will also have changes based on current `main` branch, but not as stable as official releases

**Pull request merged**
